### PR TITLE
SDIT-843 Allow updates to allocations if in different prison

### PR DESCRIPTION
**Pull request by mikehalmamoj**
But continue to prevent create of allocation in wrong prison

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/repository/OffenderProgramProfileRepository.kt [hunk 1]
```diff
@@ -11,7 +11,7 @@ interface OffenderProgramProfileRepository : JpaRepository<OffenderProgramProfil
   fun findByCourseActivityAndOffenderBooking(
     courseActivity: CourseActivity,
     offenderBooking: OffenderBooking,
-  ): OffenderProgramProfile?
+  ): List<OffenderProgramProfile>
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/repository/OffenderProgramProfileRepository.kt:14]: An offender can be allocated to a course multiple times by ending then reallocating.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/AllocationResourceIntTest.kt [hunk 2]
```diff
@@ -312,18 +312,6 @@ class AllocationResourceIntTest : IntegrationTestBase() {
           assertThat(it).contains("INVALID")
         }
     }
-
-    @Test
-    fun `should return bad request offender in wrong prison`() {
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/AllocationResourceIntTest.kt:317]: Moved under CreateAllocation tests as this now only applies when creating.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/AllocationResourceIntTest.kt [hunk 3]
```diff
@@ -344,6 +332,51 @@ class AllocationResourceIntTest : IntegrationTestBase() {
...
+      val request = upsertRequest().withBookingId(offenderAtWrongPrison.latestBooking().bookingId.toString())
+
+      upsertAllocationIsBadRequest(request)
+        .expectBody().jsonPath("userMessage").value<String> {
+          assertThat(it).contains("Prisoner is at prison=MDI, not the Course activity prison=LEI")
+        }
+    }
+
+    @Test
+    fun `should create new allocation if previously ended allocation exists`() {
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/AllocationResourceIntTest.kt:348]: Prove that we can handle multiple allocations

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/AllocationResourceIntTest.kt [hunk 4]
```diff
@@ -564,14 +608,6 @@ class AllocationResourceIntTest : IntegrationTestBase() {
         """.trimMargin(),
       )
       upsertAllocationIsOk(request)
-
-      // But un-suspending is not allowed
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/AllocationResourceIntTest.kt:568]: We trust the caller and allow any updates.

**Review by steverendell — APPROVED** [review]
lgtm

**Pull request merged**